README.md
Download YouTube Videos as MP4
===============

<h3>
===============
<a href="https://github.com/gantt/downloadyoutube/raw/master/script/yt.user.js">
    Install the script
</a> 
===============
</h3>

<br>

<div id="full_description"><b>Important</b>: If you find compatibility issues, check <a href="Debug.md">the bug reporting page</a>.
The script is also available <a href="https://greasyfork.org/scripts/1317-download-youtube-videos-as-mp4">at greasyfork.org</a>. 
<br><br><b>What's new</b>: Tampermonkey bug fix. <a href="Changelog.md">More in the changelog</a>.


<br><br><img alt="" src="http://i.imgur.com/xhdYmVg.png"><br><br>

The script works in:

<ul><li><b>Firefox</b> (install <a href="https://addons.mozilla.org/firefox/addon/greasemonkey/">Greasemonkey</a> first. Alternatively, you can install <a href="https://addons.mozilla.org/firefox/addon/download-youtube/">an extension that uses this script</a>)</li><li><b>Chrome</b> (install <a href="https://chrome.google.com/webstore/detail/dhdgffkkebhmkfjojejmpbldmpobfkfo">the Tampermonkey extension</a> first)</li><li><b>Opera</b> (install <a href="https://addons.opera.com/en/extensions/details/violent-monkey/">Violent Monkey</a> or <a href="http://addons.opera.com/extensions/details/tampermonkey-beta/">Tampermonkey</a> first. Alternatively, you can install <a href="https://addons.opera.com/addons/extensions/details/download-youtube-videos-as-mp4/">an extension which uses the script</a>)</li><li><b>Safari</b> (install the <a href="http://ss-o.net/safari/extension/NinjaKit.safariextz">NinjaKit</a> extension first. Make sure that extensions are enabled in the settings, download the file and double click it. Then you can install the script. More information <a href="http://wiki.greasespot.net/Cross-browser_userscripting#NinjaKit">here</a>.)</li></ul>


Here are the formats you can download using this script (Wikipedia has <a href="http://en.wikipedia.org/wiki/YouTube#Quality_and_codecs">the full list</a>):

<ul><li><b>MP4 360p</b> (384x288, 480x360, 640x360) - all videos are available in this format. Use it if you want to play videos on a mobile phone (iPhone, Android) or to save space.</li><li><b>MP4 720p</b> (1280x720) - HD format, available for an increasing number of videos. Use this format if you want to play videos on a tablet (iPad, Motorola Xoom, Samsung Galaxy Tab) or a laptop.</li><li><del><b>MP4 1080p</b> (1920x1080) - Full HD format, available for fewer videos than MP4 720p. Use this format to watch videos on a HD TV or a high resolution monitor.</del></li><li><del><b>MP4 4K</b> (<a href="http://youtube-global.blogspot.com/2010/07/whats-bigger-than-1080p-4k-video-comes.html">up to 4096p</a>) - HD format, available for a few videos.</del></li><li>the highest-quality FLV version available: <b>FLV 240p</b> (426x240), <del><b>FLV 360p</b> (480x360, 640x360) or <b>FLV 480p</b> (640x480, 854x480)</del>. Even if they have a different extension, .flv videos use the H.264 codec, just like .mp4 videos.</li><li>480p and 1080p are no longer available as standalone videos downloaded from youtube.com. YouTube switched to adaptive streams and uses separate files for video and audio, which need to be merged to get a video file. This is more complicated and requires more advanced tools (ffmpeg), so I haven't added them yet. </li></ul>

After downloading videos, you can play them using free applications like:
<ul><li><a href="http://www.videolan.org/vlc/">VLC</a> (Windows, Mac, Linux / plays both MP4 and FLV videos) - recommended video player</li><li><a href="http://windows.microsoft.com/en-US/windows7/products/features/windows-media-player-12">Windows Media Player 12</a> for Windows 7 (plays only MP4 videos)</li><li><a href="http://www.apple.com/itunes/download/">iTunes</a> and <a href="http://www.apple.com/quicktime/download/">QuickTime Player</a> (Windows, Mac / play only MP4 videos)</li><li>video players for mobile devices (for example, MP4 360p videos are better suited for older smartphones like iPhone 3GS, HTC Desire, Galaxy S/S2; MP4 720p videos are great for more recent smartphones (iPhone 4/4S/5, Galaxy S3), 7-inch tablets and older 10-inch tablets (iPad 1/2), while 1080p videos work better on the latest iPad, Nexus 10 and other high-resolution tablets and smartphones.)</li></ul><b>Automatic updates</b>: Scriptish and Greasemonkey for Firefox and Tampermonkey for Chrome support automatic updating.

<br><br><b>Related extensions</b>: If you use Firefox, you can also install <a href="https://addons.mozilla.org/firefox/addon/download-youtube/">an extension based on this script</a>. There's also <a href="https://addons.opera.com/addons/extensions/details/download-youtube-videos-as-mp4/">an extension for Opera</a>.</div>


Changelog.md
## Download YouTube Videos as MP4: Changelog

##### 1.8.8 (2016-09-02)
* removed outdated formats
* fixed manifest parsing (thanks, Elite2016) 

##### 1.8.7 (2016-03-19)
* [only for the script and the Firefox extension]
* replacing a function that's supposedly deprecated in Firefox 48 (thanks, Leo)

##### 1.8.6 (2016-01-18)
* [only for the script, not the extensions]
* fixes Tampermonkey cross-origin issues (thanks, Giusy)

##### 1.8.5 (2015-12-19)
* signature bug fix
* improved format detection

##### 1.8.4 (2015-11-10)
* fixed signature detection (thanks, everyone) 

##### 1.8.3 (2015-05-17)
* bug fix for experimental YouTube SPF - thanks, Alexandru 

##### 1.8.2 (2015-05-14)
* bug fix for blocked or removed videos - thanks, Сергей 
* DASH bug fixes

##### 1.8.1 (2015-03-19)
* [Firefox extension] Added support for electrolysis (e10s) - https://wiki.mozilla.org/Electrolysis
* [Firefox extension] Dropped support for Firefox 19 and earlier (e10s required a major code rewrite)
* Traditional Chinese translation - thanks, Tse Kit Yam
* Bug fixes

##### 1.7.31 (2015-01-30)
* fixes download issues

##### 1.7.30 (2014-11-14)
* fixes signature detection (thanks, everyone)

##### 1.7.29 (2014-11-01)
* fixes broken download URLs
* fixes video detection in Firefox 35+ (thanks, Nicolas)

##### 1.7.28 (2014-09-17)
* fixes missing like numbers 

##### 1.7.27 (2014-09-15)
* brings back support for the old YouTube UI

##### 1.7.26 (2014-09-03)
* added download icon and improved button integration 

##### 1.7.25 (2014-08-12)
* fix for signatures poorly detected by previous versions

##### 1.7.24
* bug fix for signature detection

##### 1.7.23
* Force refresh incorrect signatures
* Better signature detection
* Improved debugging information 

##### 1.7.22
* bug fix for signature detection (thanks, jfurg26)

##### 1.7.21
* support for YouTube's latest UI experiment (thanks, David Book)

##### 1.7.19
* bug fix for signature detection

##### 1.7.18
* Korean translation (thanks, daunmore - <a href="http://userscripts.org/topics/147092">http://userscripts.org/topics/147092</a>)
* improved URL matching
* better file size detection
* Dash improvements:
 * setting for enabling Dash formats (if you use Scriptish in Firefox or the standalone Firefox extension, go to about:config, search for download-youtube-dash-enabled and change the value to 1. To disable Dash items, change the value to 0.)
 * 256kbps Dash audio is back (thanks, Lunboks - <a href="http://userscripts.org/topics/135892">http://userscripts.org/topics/135892</a>)
 * added support for 1440p Dash videos
 * replaced 4K labeling with 2160p

##### 1.7.17
* AJAX support for channels
* dropped file size styling

##### 1.7.15
* added file size next to each download option
* bug fixes

##### 1.7.14
* removed 480p FLV, 1080p MP4, 4K MP4. They're no longer available.

##### 1.7.13
* signature bug fix

##### 1.7.12
* brings back support for 480p FLV, 1080p MP4, 4K MP4. 
Thanks, Xytras, Morten79, toejam119, ReadALot, Denisk, Tupring, milkywaif, funkinlesson and FlashGot devs.

##### 1.7.10
* fixes for the AJAX interface in Chrome

##### 1.7.9
* added low-quality FLV fallback - the 240p FLV option is added if no other FLV download URL is available
* the script now works when cookies are disabled (thanks, soft475u - <a href="http://userscripts.org/users/529340">http://userscripts.org/users/529340</a>)

##### 1.7.8
* dropped Feather support
* fix for the "share buttons" experiment - <a href="https://plus.google.com/111166827982420581810/posts/DYdHZouXbAP">https://plus.google.com/111166827982420581810/p...</a>

##### 1.7.7
* using the GreaseMonkey API to store signature data (thanks, Eric - <a href="http://userscripts.org/topics/129461">http://userscripts.org/topics/129461</a>)
* inline signature update
* Indonesian translation (thanks, <a href="http://userscripts.org/users/bayuah">http://userscripts.org/users/bayuah</a>)

##### 1.7.6
* updated Ajax support

##### 1.7.3, 1.7.5
* restored some features from 1.7

##### 1.7.2
* back to 1.6.3
* fixed signature obfuscation

##### 1.7
* fixed signature obfuscation (Firefox/Chrome)
* better support for the AJAX experiment

##### 1.6.3
* added support for Hindi (thanks, Arpit Kumar)
* fixed signature obfuscation

##### 1.6
* supports the new experimental Ajax interface
<a href="http://googlesystem.blogspot.com/2013/05/youtube-search-experiments.html">http://googlesystem.blogspot.com/2013/05/youtub...</a>
* workaround for Chrome's unsafeWindow bug fix

##### 1.5.18
* fixed the same old signature obfuscation

##### 1.5.17
* removed the Chrome workaround for unsafeWindow (thanks, ngyikp - <a href="http://userscripts.org/topics/127453">http://userscripts.org/topics/127453</a>)

##### 1.5.16
* added Turkish translation (thanks, fader ed - <a href="http://userscripts.org/users/240701">http://userscripts.org/users/240701</a>)
* cosmetic fixes for non-English interfaces
* more bug fixes

##### 1.5.10-15
* keeping up with YouTube's experiments for the signature parameter

##### 1.5.9
* Chinese and Japanese translations (thanks, SiwyPL and JayXon - <a href="http://userscripts.org/users/420869">http://userscripts.org/users/420869</a>)
* bug fix for encrypted signatures

##### 1.5.7.2
* fix for obtaining video information

##### 1.5.7.1
* keeping up with YouTube's experiments for the signature parameter

##### 1.5.7
* Polish translation (thanks, SiwyPL)
* keeping up with YouTube's experiments for the signature parameter

##### 1.5.4-1.5.6
* more bug fixes for the blank download page (YouTube experiments with new signature formats)

##### 1.5.3
* bug fix for blank download page (thanks, Caroline)

##### 1.5.1
* Feather support - <a href="http://www.youtube.com/feather_beta">http://www.youtube.com/feather_beta</a> (thanks, Joe)
* removed FLV 240p

##### 1.5
* faster video info detection
* Czech translation (thanks, Baegus - <a href="http://userscripts.org/users/495526">http://userscripts.org/users/495526</a>)

firefox/bootstrap.js
'use strict';

const{classes:Cc,interfaces:Ci,utils:Cu}=Components;
Cu.import('resource://gre/modules/Services.jsm');

var YoutubempEngine={

packageName:'downloadyoutubemp4',
frameName:'',

init:function() {
	 let mm = Cc['@mozilla.org/globalmessagemanager;1'].getService(Ci.nsIMessageListenerManager);
	 YoutubempEngine.frameName = 'chrome://'+YoutubempEngine.packageName+'/content/overlay.js'+'?' + Math.random();
	 mm.loadFrameScript(YoutubempEngine.frameName, true); // workaround for frame caching bug - https://bugzilla.mozilla.org/show_bug.cgi?id=1051238
	 mm.broadcastAsyncMessage(YoutubempEngine.packageName+':init', YoutubempEngine.packageName);
	 mm.addMessageListener(YoutubempEngine.packageName+':savefileas', YoutubempEngine.saveFileAs);	 
	 mm.addMessageListener(YoutubempEngine.packageName+':setvalue', YoutubempEngine.setValue);
},

uninit:function() {
	 let mm = Cc['@mozilla.org/globalmessagemanager;1'].getService(Ci.nsIMessageListenerManager);
	 mm.broadcastAsyncMessage(YoutubempEngine.packageName+':uninit', '');
	 mm.removeMessageListener(YoutubempEngine.packageName+':savefileas', YoutubempEngine.saveFileAs);	 
	 mm.removeMessageListener(YoutubempEngine.packageName+':setvalue', YoutubempEngine.setValue);	 
	 mm.removeDelayedFrameScript(YoutubempEngine.frameName); // frame script is not loaded into any new tabs
},

saveFileAs: function(message) {
  let url = message.data.url;
  let filename = message.data.filename;
  //let win = message.objects.target;
  let nsIFilePicker = Ci.nsIFilePicker;
  let fp = Cc['@mozilla.org/filepicker;1'].createInstance(nsIFilePicker);
  fp.defaultString = filename;
  let directoryString = YoutubempEngine.getValue({data:{pref:'download-youtube-last-directory'}});
  let directory = null;
  if (directoryString) { // last download directory
    try {
    directory = Cc['@mozilla.org/file/local;1'].
    createInstance(Ci.nsIFile).initWithPath(directoryString);
    } catch(e) { }
    if (directory === null) { // reset last download directory
      directoryString = null;
      YoutubempEngine.setValue({data:{pref:'download-youtube-last-directory', value:''}});
    }
  }
  if (directoryString === null) { // default download directory
    directory = null;
  }
  
  fp.displayDirectory = directory;
  let pos = filename.lastIndexOf('.');
  let extension = (pos>-1)?filename.substring(pos+1):'*';
  fp.appendFilter((extension=='m4a')?'Audio':'Video', '*.'+extension);
  
  var windowMediator = Cc['@mozilla.org/appshell/window-mediator;1'].
  getService(Ci.nsIWindowMediator);
  var window = windowMediator.getMostRecentWindow(null);
  
  fp.init(window, null, nsIFilePicker.modeSave);
  let fileBox = fp.show();
  try {
  if (fileBox == nsIFilePicker.returnOK || nsIFilePicker.returnReplace) {     
     let ioService = Cc['@mozilla.org/network/io-service;1'].
     getService(Ci.nsIIOService);
     let uri = ioService.newURI(url, null , null);
     let fileURI = ioService.newFileURI(fp.file);
     let persist = Cc['@mozilla.org/embedding/browser/nsWebBrowserPersist;1'].createInstance(Ci.nsIWebBrowserPersist);       
     let xfer = Cc['@mozilla.org/transfer;1'].createInstance(Ci.nsITransfer);
     let privacyContext = window.QueryInterface(Ci.nsIInterfaceRequestor).getInterface(Ci.nsIWebNavigation).QueryInterface(Ci.nsILoadContext);
     xfer.init(uri, fileURI, '', null, null, null, persist, false);  
     persist.progressListener = xfer;
     try { // Firefox 36+ (new parameter in Firefox 36: aReferrerPolicy)
      persist.saveURI(uri, null, null, 0, null, null, fp.file, privacyContext);
     } catch(e) { // Firefox <36 -> only 7 parameters
      persist.saveURI(uri, null, null, null, null, fp.file, privacyContext);
     }               
     if (fp.file.parent.path != directoryString) {
        YoutubempEngine.setValue({data:{pref:'download-youtube-last-directory', value:fp.file.parent.path}});
     }
   }
   } catch(e) { }
},

getValue: function(message) {
  let pref = message.data.pref;
  let settings = Cc['@mozilla.org/preferences-service;1'].
    getService(Ci.nsIPrefService).
    getBranch('extensions.'+YoutubempEngine.packageName+'.');
  let prefType = settings.getPrefType(pref);
  return (prefType == settings.PREF_STRING)?settings.getCharPref(pref):null;
},

setValue: function(message) {
  let pref = message.data.pref;
  let value = message.data.value;
  let settings = Cc['@mozilla.org/preferences-service;1'].
    getService(Ci.nsIPrefService).
    getBranch('extensions.'+YoutubempEngine.packageName+'.');  
  settings.setCharPref(pref, value);
}
}

function startup(data,reason) {
  let alias = Services.io.newFileURI(data.installPath);
  if (!data.installPath.isDirectory()) {
    alias = Services.io.newURI('jar:' + alias.spec + '!/', null, null);
  }
  // resource registration
  let resource = Services.io.getProtocolHandler('resource').QueryInterface(Ci.nsIResProtocolHandler);
  resource.setSubstitution(YoutubempEngine.packageName, alias);  
	YoutubempEngine.init();
}

function shutdown(data,reason) {
  // remove resource registration
  let resource = Services.io.getProtocolHandler('resource').QueryInterface(Ci.nsIResProtocolHandler);
  resource.setSubstitution(YoutubempEngine.packageName, null);
	YoutubempEngine.uninit();
}

function install(data,reason) {}
function uninstall(data,reason) {}


operapresto/scripts/background.js
opera.extension.onmessage = function(event) {
  var action = event.data.action;
  //opera.postError(action);
  
  xhrequest=function(event) {
    var xhr = new XMLHttpRequest();
    xhr.open(event.data.method, event.data.url, false);
    xhr.onreadystatechange = function() {
      if (this.readyState == 4) {
        if (this.status >= 301 && this.status < 400) { //redirects 
          var location=this.getResponseHeader('Location');
          if (location) {
            //opera.postError(this.status);
            //opera.postError(this.getAllResponseHeaders());
            event.data.url=location;
            xhrequest(event);
            return;
          }
        }
        event.source.postMessage({
          action:action+'-response', 
          responseText:this.responseText||'', 
          readyState:this.readyState, 
          status:this.status, 
          responseHeaders:this.getAllResponseHeaders()
        });
        //opera.postError(this.status);
        //opera.postError(event.data.url);
        //opera.postError(this.getAllResponseHeaders());
      }
    };
    xhr.send();  
  }
  
  if (action.indexOf('xhr') == 0) {
    xhrequest(event);
  }
};


firefox/content/overlay.js
'use strict';

const{classes:Cc,interfaces:Ci,utils:Cu}=Components;

let YoutubempEngine={

packageName: 'downloadyoutubemp4',
safeWin: null,
target: null,

dataStorage: function () { // GM_getValue, GM_setValue
  this.pref = Cc['@mozilla.org/preferences-service;1'].
    getService(Ci.nsIPrefService).
    getBranch('extensions.'+YoutubempEngine.packageName+'.');    
  this.getValue = function(prefName, defaultValue) {
    let prefType = this.pref.getPrefType(prefName);
    return (prefType == this.pref.PREF_STRING)?this.pref.getCharPref(prefName):defaultValue;
  }
  this.setValue = function(prefName, value) {
    if (typeof value == 'string') {
      sendAsyncMessage(YoutubempEngine.packageName+':setvalue', {pref:prefName,value:value});
    }
  }  
},

crossXHR: function (unsafeContentWin) { // GM_xmlhttpRequest
  this.unsafeContentWin = unsafeContentWin;
  this.contentStartRequest = function(details) {
    if (Cu.waiveXrays) { // bypass xrays protection Firefox 32+
      details = Cu.waiveXrays(details);
    }    
    let url = details.url;
    if (typeof url != 'string' || !/^https?:\/\//.test(url)) {
      throw new Error('crossXHR: Invalid url '+url);
    }
    new XPCNativeWrapper(unsafeContentWin, 'setTimeout()')
    .setTimeout(YoutubempEngine.hitch(this, 'chromeStartRequest', details), 0);
  }
  this.chromeStartRequest = function(details) {
    let req = Cc['@mozilla.org/xmlextras/xmlhttprequest;1']
    .createInstance(Ci.nsIXMLHttpRequest);
    this.setupRequestEvent(this.unsafeContentWin, req, 'load', details);
    this.setupRequestEvent(this.unsafeContentWin, req, 'error', details);
    this.setupRequestEvent(this.unsafeContentWin, req, 'readystatechange', details);
    req.open(details.method, details.url);
    req.send(details.data);
  }
  this.setupRequestEvent = function(unsafeContentWin, req, event, details) {
    if (!details['on' + event]) return; 
    req.addEventListener(event, function(evt) {
       let responseState = {
       // __exposedProps__ is deprecated, but it's still used for Firefox < 35
          __exposedProps__: {readyState:'r',responseText:'r',
          responseHeaders:'r',status:'r',statusText:'r'},       
          responseText: req.responseText,
          responseHeaders: req.getAllResponseHeaders(),
          readyState: req.readyState,
          status: (req.readyState==4)?req.status:0,
          statusText: (req.readyState==4)?req.statusText:''
        };
        let state = responseState;
        // __exposedProps__ alternative, cloneInto requires Firefox 35+
        if (typeof Cu.cloneInto == 'function') {
          state = Cu.cloneInto(responseState,unsafeContentWin);
        }
        new XPCNativeWrapper(unsafeContentWin, 'setTimeout()')
          .setTimeout(function(){ details['on' + event](state); }, 0);
    }, false);
  }  
},
  
hitch: function (obj, method) {
  if (obj && method && (typeof method == 'string')) {
    if (!obj[method]) {
      throw 'Hitch: '+obj+'.'+method+' does not exist';
    }
    method = obj[method];
  } else if (typeof method == 'function') {
    obj = obj || {};
  } else {
    throw 'Hitch: Invalid arguments';
  }
  let staticArgs = Array.prototype.splice.call(arguments, 2, arguments.length);
  return function() {
    let args = Array.prototype.slice.call(staticArgs);
    Array.prototype.push.apply(args, arguments);
    return method.apply(obj, args);
  };
},

downloadFile: function(url, filename) {
    sendAsyncMessage(YoutubempEngine.packageName+':savefileas', 
    {url:url,filename:filename}, {target:YoutubempEngine.target});
},

getUrlContents: function(url) {
    let	ioService = Cc['@mozilla.org/network/io-service;1']
    .getService(Ci.nsIIOService);
    let	scriptableStream = Components
    .classes['@mozilla.org/scriptableinputstream;1']
    .getService(Ci.nsIScriptableInputStream);
    let unicodeConverter = Components
    .classes['@mozilla.org/intl/scriptableunicodeconverter']
    .createInstance(Ci.nsIScriptableUnicodeConverter);
    unicodeConverter.charset = 'UTF-8';
    let input = null;
    if (typeof ioService.newChannel == 'function') { // deprecated in Firefox 48
    	input = ioService.newChannel(url, 'UTF-8', null).open();
    } else {
    	input = ioService.newChannel2(url, 'UTF-8', null, null, null, null, 
    	Ci.nsILoadInfo.SEC_NORMAL, 
    	Ci.nsIContentPolicy.TYPE_OTHER).open();
    }    
    scriptableStream.init(input);
    let	str = scriptableStream.read(input.available());
    scriptableStream.close();
    input.close();
    try {
      return unicodeConverter.ConvertToUnicode(str);
    } catch (e) {
      return str;
    }
},

contentLoad: function(e) {
    YoutubempEngine.target = e.target;
    let unsafeWin = e.target.defaultView;
    if (unsafeWin.wrappedJSObject) {
      unsafeWin = unsafeWin.wrappedJSObject;
    }
    if (unsafeWin.frameElement) { // ignore frames
      return;
    }      
    let unsafeLoc = new XPCNativeWrapper(unsafeWin, 'location').location;
    let href = new XPCNativeWrapper(unsafeLoc, 'href').href;
    let scheme = Cc['@mozilla.org/network/io-service;1']
    .getService(Ci.nsIIOService).extractScheme(href);   
    if ((scheme == 'http' || scheme == 'https') && 
    /^https?:\/\/www\.youtube\.com\//.test(href) && 
    !/^https?:\/\/www\.youtube\.com\/embed\//.test(href)) { // inject script
      let safeWin = new XPCNativeWrapper(unsafeWin);
      YoutubempEngine.safeWin = e.target.defaultView;
      let sandbox = new Cu.Sandbox(safeWin, 
      {'sandboxPrototype':safeWin, 'wantXrays':true});
      var unsafeWindowGetter = new sandbox.Function('return window.wrappedJSObject || window;');
      Object.defineProperty(sandbox, 'unsafeWindow', {get: unsafeWindowGetter});
      let storage = new YoutubempEngine.dataStorage();
      sandbox.GM_getValue = YoutubempEngine.hitch(storage, 'getValue');
      sandbox.GM_setValue = YoutubempEngine.hitch(storage, 'setValue');
      let xmlhttpRequester = new YoutubempEngine.crossXHR(unsafeWin);
      sandbox.GM_xmlhttpRequest = YoutubempEngine.hitch(xmlhttpRequester, 'contentStartRequest');      
      sandbox.GM_download = YoutubempEngine.downloadFile;
      try {
          let script = YoutubempEngine.getUrlContents(
          'resource://'+YoutubempEngine.packageName+'/content/yt.user.js');
          Cu.evalInSandbox(script, sandbox);
      } catch (e) { }
    }       
},

inject: function(event) {
  let doc = event.originalTarget;
  if (doc.nodeName != '#document') return; // only documents 
  YoutubempEngine.contentLoad(event);
  //removeEventListener('DOMContentLoaded', YoutubempEngine.inject);
},

init: function() {
  addEventListener('DOMContentLoaded', YoutubempEngine.inject);
},

uninit: function() {
  removeEventListener('DOMContentLoaded', YoutubempEngine.inject);
}

}

// addMessageListener(YoutubempEngine.packageName+':init', YoutubempEngine.init);
// addMessageListener(YoutubempEngine.packageName+':uninit', YoutubempEngine.uninit);

addEventListener('DOMContentLoaded', YoutubempEngine.inject);